File: src/math/vector3.cpp
/*
This file is a part of stakeforge_engine: https://github.com/inanevin/stakeforge
Copyright [2025-] Inan Evin

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice, this
	  list of conditions and the following disclaimer.

   2. Redistributions in binary form must reproduce the above copyright notice,
	  this list of conditions and the following disclaimer in the documentation
	  and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "vector3.hpp"
#include "math.hpp"
#include "math/easing.hpp"
#include "data/ostream.hpp"
#include "data/istream.hpp"

#ifdef SFG_JSON_SERIALIZE
#include "vendor/nhlohmann/json.hpp"
#endif

namespace SFG
{
	const vec3f vec3f::zero(0.0f, 0.0f, 0.0f);
	const vec3f vec3f::one(1.0f, 1.0f, 1.0f);
	const vec3f vec3f::up(0.0f, 1.0f, 0.0f);
	const vec3f vec3f::forward(0.0f, 0.0f, -1.0f);
	const vec3f vec3f::right(1.0f, 0.0f, 0.0f);

	vec3f vec3f::clamp(const vec3f& vector, const vec3f& min_vec, const vec3f& max_vec)
	{
		return vec3f(math::clamp(vector.x, min_vec.x, max_vec.x), math::clamp(vector.y, min_vec.y, max_vec.y), math::clamp(vector.z, min_vec.z, max_vec.z));
	}

	vec3f vec3f::cross(const vec3f& a, const vec3f& b)
	{
		return vec3f(a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x);
	}

	vec3f vec3f::abs(const vec3f& vector)
	{
		return vec3f(math::abs(vector.x), math::abs(vector.y), math::abs(vector.z));
	}

	vec3f vec3f::min(const vec3f& a, const vec3f& b)
	{
		return vec3f(math::min(a.x, b.x), math::min(a.y, b.y), math::min(a.z, b.z));
	}

	vec3f vec3f::max(const vec3f& a, const vec3f& b)
	{
		return vec3f(math::max(a.x, b.x), math::max(a.y, b.y), math::max(a.z, b.z));
	}

	vec3f vec3f::lerp(const vec3f& a, const vec3f& b, f32 t)
	{
		return vec3f(easing::lerp(a.x, b.x, t), easing::lerp(a.y, b.y, t), easing::lerp(a.z, b.z, t));
	}

	f32 vec3f::dot(const vec3f& a, const vec3f& b)
	{
		return a.x * b.x + a.y * b.y + a.z * b.z;
	}

	f32 vec3f::distance(const vec3f& a, const vec3f& b)
	{
		return (a - b).magnitude();
	}

	f32 vec3f::distance_sqr(const vec3f& a, const vec3f& b)
	{
		return (a - b).magnitude_sqr();
	}

	vec3f vec3f::project(const vec3f& on_normal) const
	{
		return on_normal * dot(*this, on_normal);
	}

	vec3f vec3f::rotate(const vec3f& axis, f32 angle_degrees) const
	{
		vec3f unit_axis = axis.normalized();
		if (unit_axis.is_zero())
		{
			return *this;
		}

		f32 angle_rad = math::degrees_to_radians(angle_degrees);
		f32 cos_theta = math::cos(angle_rad);
		f32 sin_theta = math::sin(angle_rad);

		vec3f v_rot = (*this * cos_theta) + (vec3f::cross(unit_axis, *this) * sin_theta) + (unit_axis * (vec3f::dot(unit_axis, *this) * (1.0f - cos_theta)));
		return v_rot;
	}

	vec3f vec3f::reflect(const vec3f& in_normal) const
	{
		vec3f unit_normal = in_normal.normalized();
		if (unit_normal.is_zero())
		{
			return -(*this);
		}
		return *this - (unit_normal * (2.0f * vec3f::dot(*this, unit_normal)));
	}

	bool vec3f::equals(const vec3f& other, f32 epsilon) const
	{
		return math::almost_equal(x, other.x, epsilon) && math::almost_equal(y, other.y, epsilon) && math::almost_equal(z, other.z, epsilon);
	}

	bool vec3f::is_zero(f32 epsilon) const
	{
		return math::almost_equal(x, 0.0f, epsilon) && math::almost_equal(y, 0.0f, epsilon) && math::almost_equal(z, 0.0f, epsilon);
	}

	f32 vec3f::magnitude() const
	{
		return math::sqrt(x * x + y * y + z * z);
	}

	f32 vec3f::magnitude_sqr() const
	{
		return x * x + y * y + z * z;
	}

	void vec3f::serialize(ostream& stream) const
	{
		stream << x << y << z;
	}

	void vec3f::deserialize(istream& stream)
	{
		stream >> x >> y >> z;
	}

#ifdef SFG_JSON_SERIALIZE
	void to_json(nlohmann::json& j, const vec3f& v)
	{
		j = nlohmann::json::array({v.x, v.y, v.z});
	}

	void from_json(const nlohmann::json& j, vec3f& v)
	{
		if (!j.is_array() || j.size() < 3)
			throw std::runtime_error("vec3f json err");

		v.x = j.at(0).get<f32>();
		v.y = j.at(1).get<f32>();
		v.z = j.at(2).get<f32>();
	}

#endif
}

File: stakeforge_managed/Stakeforge.Managed/Vector3.cs
using System;
using System.Runtime.InteropServices;

namespace SFG;

[StructLayout(LayoutKind.Sequential)]
public struct Vector3 : IEquatable<Vector3>
{
    public static Vector3 Zero => new(0.0f, 0.0f, 0.0f);
    public static Vector3 One => new(1.0f, 1.0f, 1.0f);
    public static Vector3 Up => new(0.0f, 1.0f, 0.0f);
    public static Vector3 Down => new(0.0f, -1.0f, 0.0f);
    public static Vector3 Forward => new(0.0f, 0.0f, -1.0f);
    public static Vector3 Back => new(0.0f, 0.0f, 1.0f);
    public static Vector3 Right => new(1.0f, 0.0f, 0.0f);
    public static Vector3 Left => new(-1.0f, 0.0f, 0.0f);

    public float X;
    public float Y;
    public float Z;

    public Vector3(float x, float y, float z)
    {
        X = x;
        Y = y;
        Z = z;
    }

    public readonly float Length => Math.Sqrt(LengthSquared);
    public readonly float LengthSquared => X * X + Y * Y + Z * Z;
    public readonly float Magnitude => Length;
    public readonly float MagnitudeSquared => LengthSquared;

    public readonly Vector3 Normalized
    {
        get
        {
            float length = Length;
            return length > Math.Epsilon ? this / length : Zero;
        }
    }

    public void Normalize()
    {
        this = Normalized;
    }

    public static Vector3 Clamp(Vector3 value, Vector3 minimum, Vector3 maximum)
    {
        return new Vector3(Math.Clamp(value.X, minimum.X, maximum.X), Math.Clamp(value.Y, minimum.Y, maximum.Y), Math.Clamp(value.Z, minimum.Z, maximum.Z));
    }

    public static Vector3 ClampMagnitude(Vector3 value, float maximumLength)
    {
        float length = value.Length;
        return length > maximumLength ? value / length * maximumLength : value;
    }

    public static Vector3 Cross(Vector3 a, Vector3 b)
    {
        return new Vector3(a.Y * b.Z - a.Z * b.Y, a.Z * b.X - a.X * b.Z, a.X * b.Y - a.Y * b.X);
    }

    public static Vector3 Abs(Vector3 value) => new(Math.Abs(value.X), Math.Abs(value.Y), Math.Abs(value.Z));
    public static Vector3 Min(Vector3 a, Vector3 b) => new(Math.Min(a.X, b.X), Math.Min(a.Y, b.Y), Math.Min(a.Z, b.Z));
    public static Vector3 Max(Vector3 a, Vector3 b) => new(Math.Max(a.X, b.X), Math.Max(a.Y, b.Y), Math.Max(a.Z, b.Z));
    public static Vector3 Lerp(Vector3 a, Vector3 b, float amount) => new(Math.Lerp(a.X, b.X, amount), Math.Lerp(a.Y, b.Y, amount), Math.Lerp(a.Z, b.Z, amount));
    public static float Dot(Vector3 a, Vector3 b) => a.X * b.X + a.Y * b.Y + a.Z * b.Z;
    public static float Distance(Vector3 a, Vector3 b) => (a - b).Length;
    public static float DistanceSquared(Vector3 a, Vector3 b) => (a - b).LengthSquared;

    public static Vector3 Project(Vector3 value, Vector3 onNormal)
    {
        float denominator = onNormal.LengthSquared;
        return denominator < Math.Epsilon ? Zero : onNormal * (Dot(value, onNormal) / denominator);
    }

    public static Vector3 Reflect(Vector3 value, Vector3 normal)
    {
        Vector3 unitNormal = normal.Normalized;
        return unitNormal.IsZero() ? -value : value - unitNormal * (2.0f * Dot(value, unitNormal));
    }

    public static Vector3 Rotate(Vector3 value, Vector3 axis, float angleDegrees)
    {
        Vector3 unitAxis = axis.Normalized;

        if (unitAxis.IsZero())
        {
            return value;
        }

        float radians = Math.DegreesToRadians(angleDegrees);
        float cosine = Math.Cos(radians);
        float sine = Math.Sin(radians);
        return value * cosine + Cross(unitAxis, value) * sine + unitAxis * (Dot(unitAxis, value) * (1.0f - cosine));
    }

    public readonly bool ApproximatelyEquals(Vector3 other, float epsilon = Math.Epsilon)
    {
        return Math.ApproximatelyEqual(X, other.X, epsilon) && Math.ApproximatelyEqual(Y, other.Y, epsilon) && Math.ApproximatelyEqual(Z, other.Z, epsilon);
    }

    public readonly bool IsZero(float epsilon = Math.Epsilon)
    {
        return Math.Abs(X) < epsilon && Math.Abs(Y) < epsilon && Math.Abs(Z) < epsilon;
    }

    public readonly bool Equals(Vector3 other) => X.Equals(other.X) && Y.Equals(other.Y) && Z.Equals(other.Z);
    public override readonly bool Equals(object? obj) => obj is Vector3 other && Equals(other);
    public override readonly int GetHashCode() => HashCode.Combine(X, Y, Z);
    public override readonly string ToString() => $"({X}, {Y}, {Z})";

    public static Vector3 operator +(Vector3 a, Vector3 b) => new(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
    public static Vector3 operator -(Vector3 a, Vector3 b) => new(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
    public static Vector3 operator -(Vector3 value) => new(-value.X, -value.Y, -value.Z);
    public static Vector3 operator *(Vector3 value, float scalar) => new(value.X * scalar, value.Y * scalar, value.Z * scalar);
    public static Vector3 operator *(float scalar, Vector3 value) => value * scalar;
    public static Vector3 operator *(Vector3 a, Vector3 b) => new(a.X * b.X, a.Y * b.Y, a.Z * b.Z);
    public static Vector3 operator /(Vector3 value, float scalar) => scalar == 0.0f ? Zero : new(value.X / scalar, value.Y / scalar, value.Z / scalar);
    public static bool operator ==(Vector3 a, Vector3 b) => a.Equals(b);
    public static bool operator !=(Vector3 a, Vector3 b) => !a.Equals(b);
}
